File: lua/day14_part2_2015.lua

function readReindeerDetails(filename)
    local file = io.open(filename, "r")
    local reindeers = {}
    
    for line in file:lines() do
        local parts = {}
        for part in line:gmatch("%S+") do
            table.insert(parts, part)
        end
        local speed = tonumber(parts[4])
        local flyTime = tonumber(parts[7])
        local restTime = tonumber(parts[14])
        
        table.insert(reindeers, {speed = speed, flyTime = flyTime, restTime = restTime, flying = true, distance = 0, points = 0, timeInMode = 0})
    end
    
    file:close()
    return reindeers
end

function simulateRaceWithPoints(reindeers, totalSeconds)
    for i = 1, totalSeconds do
        local maxDistance = 0
        for j, reindeer in ipairs(reindeers) do
            if reindeer.flying then
                reindeer.distance = reindeer.distance + reindeer.speed
            end
            reindeer.timeInMode = reindeer.timeInMode + 1
            if (reindeer.flying and reindeer.timeInMode == reindeer.flyTime) or (not reindeer.flying and reindeer.timeInMode == reindeer.restTime) then
                reindeer.flying = not reindeer.flying
                reindeer.timeInMode = 0
            end
            if reindeer.distance > maxDistance then
                maxDistance = reindeer.distance
            end
        end
        for j, reindeer in ipairs(reindeers) do
            if reindeer.distance == maxDistance then
                reindeer.points = reindeer.points + 1
            end
        end
    end
end

function findMaxPoints(reindeers)
    local maxPoints = 0
    for _, reindeer in ipairs(reindeers) do
        if reindeer.points > maxPoints then
            maxPoints = reindeer.points
        end
    end
    return maxPoints
end

local reindeers = readReindeerDetails("input.txt")
simulateRaceWithPoints(reindeers, 2503)
local maxPoints = findMaxPoints(reindeers)
print(maxPoints)

File: r/day14_part2_2015.R
read_reindeer_details <- function(filename) {
  reindeers <- list()
  lines <- readLines(filename)
  
  for (line in lines) {
    parts <- unlist(strsplit(line, " "))
    speed <- as.integer(parts[4])
    fly_time <- as.integer(parts[7])
    rest_time <- as.integer(parts[14])
    reindeers <- append(reindeers, list(list(speed = speed, fly_time = fly_time, rest_time = rest_time, distance = 0, points = 0, flying = TRUE, time_in_mode = 0)))
  }
  
  return(reindeers)
}

simulate_race_with_points <- function(reindeers, total_seconds) {
  for (i in 1:total_seconds) {
    max_distance <- 0
    for (j in seq_along(reindeers)) {
      reindeer <- reindeers[[j]]
      if (reindeer$flying) reindeer$distance <- reindeer$distance + reindeer$speed
      reindeer$time_in_mode <- reindeer$time_in_mode + 1
      if (reindeer$flying && reindeer$time_in_mode == reindeer$fly_time || !reindeer$flying && reindeer$time_in_mode == reindeer$rest_time) {
        reindeer$flying <- !reindeer$flying
        reindeer$time_in_mode <- 0
      }
      reindeers[[j]] <- reindeer
      if (reindeer$distance > max_distance) max_distance <- reindeer$distance
    }
    for (j in seq_along(reindeers)) {
      if (reindeers[[j]]$distance == max_distance) reindeers[[j]]$points <- reindeers[[j]]$points + 1
    }
  }
  return(reindeers)
}

find_max_points <- function(reindeers) {
  max_points <- 0
  for (reindeer in reindeers) {
    if (reindeer$points > max_points) max_points <- reindeer$points
  }
  return(max_points)
}

main <- function() {
  reindeers <- read_reindeer_details("input.txt")
  reindeers <- simulate_race_with_points(reindeers, 2503)
  max_points <- find_max_points(reindeers)
  print(max_points)
}

main()